File: Python/ds/Double_ended_queue.py
'''
Flexible Dequeue:
we can use both the front and rear end for insertion and deletion operations
'''


class Deque:

    # constructor function called during object creation
    def __init__(self):
        self.ir = 0
        self.il = - 1
        self.dl = 0
        self.dr = 0
        self.list = []
        self.n = 0

    # resets the queue by values entered by user at the time of creation
    def reset(self):
        self.ir = self.n
        self.il = - 1
        self.dl = 0
        self.dr = self.n-1

    # takes the length of elements to be inserted
    def entry(self):
        self.list = []
        self.n = int(input('Enter no. of elements:'))
        self.reset()
        for i in range(self.n):
            self.list.append('-')

    # insertion from left or right
    def insert(self):
        sub = self.ir-self.il
        if sub == 1:
            print('-'*10)
            print('Queue is Full')
        elif sub == 2:
            self.il += 1
            self.list[self.il] = int(input('Enter your element: '))
        else:
            ch = input('Want to insert from left or right?(L/R): ')
            if ch == 'r' or ch == 'R':
                self.ir -= 1
                self.list[self.ir] = int(input('Enter your element:'))
            elif ch == 'l' or ch == 'L':
                self.il += 1
                self.list[self.il] = int(input('Enter your element:'))
            else:
                print('-' * 10)
                print('Invalid Choice !!!!!')

    # deletion from left or right
    def delete(self):
        if self.ir == self.n and self.il == -1:
            print('Queue is Empty')
        elif self.dr == self.dl:
            print('Deleted:', self.list[self.dr])
            self.reset()
        else:
            ch = input('Want to delete from left or right?(L/R):')
            if ch == 'l' or ch == 'L':
                print('Deleted:', self.list[self.dl])
                self.dl += 1
            elif ch == 'r' or ch == 'R':
                print('Deleted:', self.list[self.dr])
                self.dr -= 1
            else:
                print('-' * 10)
                print('Invalid Choice!!!!!!')

    # print all the elements from left or right
    def display(self):
        if self.ir == self.n and self.il == -1:
            print('Queue is Empty')
        elif self.dr == self.dl:
            print('Printing the Queue: ')
            print(self.list[self.dr])
        else:
            ch = input('Want to display from left or right?(R/L):')
            if ch == 'l' or ch == 'L':
                print ('Displaying the Queue from left: ')
                for i in range(self.dl, self.dr+1, +1):
                    print(self.list[i], end='')
            elif ch == 'r' or ch == 'R':
                print('Displaying the Queue from right: ')
                for i in range(self.dr, self.dl-1, -1):
                    print(self.list[i], end=' ')
            else:
                print('Invalid Choice!!!!!!')

    # creation of the deque
    def choose(self):
        while True:
            print('''
            PRESS  :  TO
            ----------------------
            1    :  Entry
            2    :  Insert
            3    :  Delete
            4    :  Display
            5    :  Exit
            ----------------------
            ''')

            ch = int(input('Enter your choice:'))
            if ch == 1:
                self.entry()
            elif ch == 2:
                self.insert()
            elif ch == 3:
                self.delete()
            elif ch == 4:
                self.display()
            else:
                break

ob = Deque()
ob.choose()

File: C/ds/Double_Ended_Queue.c
/*  DOUBLE ENDED QUEUE
    Double ended queue or dequeue is a special type of queue 
    wherein elements can be inserted or deleted from both ends
    i.e front and rear
*/

#include <stdio.h>
//Declare dequeue with an array arr, front and rear as members
typedef struct
{
    int arr[10];
    int front, rear;
} dequeue;
dequeue dq;

//to insert an element to the front of double ended queue
void enqueue_front(int n)
{
    //If dequeue is not full
    if ((dq.rear + 1) % 10 != dq.front)
    {
        //If dequeue is empty insert as first element
        if (dq.front == -1)
        {
            dq.front = 0;
            dq.rear = 0;
        }
        //decrement front in circular manner and insert
        else
            dq.front = (dq.front + 9) % 10;
        dq.arr[dq.front] = n;
    }
}

//to insert an element to the rear of double ended queue
void enqueue_rear(int n)
{
    if ((dq.rear + 1) % 10 != dq.front)
    {
        if (dq.rear == -1)
        {
            dq.front = 0;
            dq.rear = 0;
        }
        //increment rear in circular manner and insert
        else
            dq.rear = (dq.rear + 1) % 10;
        dq.arr[dq.rear] = n;
    }
}

//to delete an element and return it from the front of dequeue
int delete_front()
{
    int item;
    //If dequeue is empty
    if (dq.front == -1)
        return -1;
    else
    {
        item = dq.arr[dq.front];
        //if there is only one element
        if (dq.front == dq.rear)
        {
            dq.front = -1;
            dq.rear = -1;
        }
        else
            dq.front = (dq.front + 1) % 10;
    }
    return item;
}

//to delete an element and return it from the rear of dequeue
int delete_rear()
{
    int item;
    if (dq.rear == -1)
        return -1;
    else
    {
        item = dq.arr[dq.rear];
        if (dq.front == dq.rear)
        {
            dq.front = -1;
            dq.rear = -1;
        }
        else
            dq.rear = (dq.rear + 9)%10;
    }
    return item;
}

//to display the double ended queue from front to rear
void display()
{
    if (dq.front <= dq.rear)
    {
        for (int i = dq.front; i <= dq.rear; i++)
            printf("%d ", dq.arr[i]);
    }
    else
    {
        for (int i = dq.front; i <= 9; i++)
            printf("%d ", dq.arr[i]);
        for (int i = 0; i <= dq.rear; i++)
            printf("%d ", dq.arr[i]);
    }
    printf("\n");
}

//driver code
int main()
{
    int q, t, n;
    //input no of queries
    scanf("%d", &q);
    dq.front = -1;
    dq.rear = -1;

    //menu driven program to insert,delete or display dequeue
    while (q--)
    {
        scanf("%d", &t);
        switch (t)
        {
        case 1:
            scanf("%d", &n);
            enqueue_front(n);
            break;
        case 2:
            scanf("%d", &n);
            enqueue_rear(n);
            break;
        case 3:
            printf("%d\n", delete_front());
            break;
        case 4:
            printf("%d\n", delete_rear());
            break;
        case 5:
            display();
            break;
        }
    }
    return 0;
}
/* 
    Sample Input:
    7
    1
    2
    2
    1
    1
    3
    1
    4
    2
    5
    3
    5
    Sample Output:
    4
    3 2 1 5
*/
/*
    Time Complexity of insertion and deletion operations: O(1)
    Time Complexity of display function : O(n)
    Time complexity of random access by index : O(1)
    Space complexity: O(1)
*/
